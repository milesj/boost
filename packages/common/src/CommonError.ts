import { createScopedError } from '@boost/internal';

const errors = {
	PARSE_INVALID_EXT: 'Unable to parse file "{0}". Unsupported file extension.',
	PATH_REQUIRE_ABSOLUTE: 'An absolute file path is required.',
	PATH_RESOLVE_LOOKUPS: 'Failed to resolve a path using the following lookups (in order):\n{0}\n',
	PROJECT_NO_PACKAGE: 'No `package.json` found within project root.',
};

export type CommonErrorCode = keyof typeof errors;

export default createScopedError<CommonErrorCode>('CMN', 'CommonError', errors);
