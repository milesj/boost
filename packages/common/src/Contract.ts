import optimal, { Blueprint, Predicates, predicates } from 'optimal';
import { Optionable } from './types';

export default abstract class Contract<T extends object = {}> implements Optionable<T> {
  readonly options: Readonly<Required<T>>;

  constructor(options?: T) {
    this.options = this.configure(options);
  }

  /**
   * Set an options object by merging the new partial and existing options
   * with the defined blueprint, while running all validation checks.
   * Freeze and return the options object.
   */
  configure(options?: Partial<T> | ((options: Required<T>) => Partial<T>)): Readonly<Required<T>> {
    const nextOptions = typeof options === 'function' ? options(this.options) : options;

    // We don't want the options property to be modified directly,
    // so it's read only, but we still want to modify it with this function.
    // @ts-expect-error
    this.options = Object.freeze(
      optimal(
        { ...this.options, ...nextOptions },
        this.blueprint(predicates, this.options === undefined) as Blueprint<T>,
        {
          name: this.constructor.name,
        },
      ),
    );

    return this.options;
  }

  /**
   * Define an optimal blueprint in which to validate and build the
   * options object passed to the constructor, or when manual setting.
   */
  abstract blueprint(predicates: Predicates, onConstruction?: boolean): Blueprint<object>;
}
