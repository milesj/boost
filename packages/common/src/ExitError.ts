export class ExitError extends Error {
	code: number;

	constructor(message: string, code: number) {
		super(message);

		this.code = code;
		this.name = 'ExitError';
	}
}
