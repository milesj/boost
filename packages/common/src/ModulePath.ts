import path from 'path';
import { ModuleID, Pathable, PortablePath } from './types';

/**
 * An immutable class for operating on Node.js module IDs, names, and paths.
 */
export class ModulePath implements Pathable {
	protected internalPath: string = '';

	protected isNormalized: boolean = false;

	constructor(...parts: PortablePath[]) {
		this.internalPath = path.join(...parts.map(String));
	}

	/**
	 * Create and return a new `ModulePath` instance.
	 */
	static create(id: PortablePath): ModulePath {
		return new ModulePath(id);
	}

	/**
	 * Append path parts to the end of the current path
	 * and return a new `ModulePath` instance.
	 */
	append(...parts: PortablePath[]): ModulePath {
		return new ModulePath(this.internalPath, ...parts);
	}

	/**
	 * Return true if the module is scoped within a private namespace
	 * (starts with @).
	 *
	 */
	hasScope(): boolean {
		return this.internalPath.startsWith('@');
	}

	/**
	 * Return the module name without any trailing import paths,
	 * or optionally without the private scope.
	 */
	name(withoutScope: boolean = false): string {
		const parts = this.path().split('/');

		if (this.hasScope() && withoutScope) {
			return parts[1];
		}

		return parts.slice(0, this.hasScope() ? 2 : 1).join('/');
	}

	/**
	 * Return the current module path as a normalized string.
	 */
	path(): ModuleID {
		if (!this.isNormalized) {
			this.isNormalized = true;
			this.internalPath = path
				.normalize(this.internalPath)
				// Node modules must always use forward slashes
				.replace(/\\/g, '/');
		}

		return this.internalPath;
	}

	/**
	 * Return the private scope with leading @, or null if not defined.
	 */
	scope(): string | null {
		if (!this.hasScope()) {
			return null;
		}

		return this.path().slice(0, this.path().indexOf('/'));
	}

	toJSON(): ModuleID {
		return this.path();
	}

	toString(): ModuleID {
		return this.path();
	}
}
