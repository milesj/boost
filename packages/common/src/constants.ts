// https://github.com/npm/validate-npm-package-name
export const MODULE_NAME_PART = /[a-z0-9]{1}[-a-z0-9_.]*/u;

export const MODULE_NAME_PATTERN = new RegExp(
	`^(?:@(${MODULE_NAME_PART.source})/)?(${MODULE_NAME_PART.source})$`,
	'u',
);
