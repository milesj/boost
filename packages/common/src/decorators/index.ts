export { default as Bind } from './Bind';
export { default as Debounce } from './Debounce';
export { default as Deprecate } from './Deprecate';
