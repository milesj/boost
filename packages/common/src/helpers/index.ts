export * from './createBlueprint';
export * from './deepFreeze';
export * from './deepMerge';
export * from './formatMs';
export * from './instanceOf';
export * from './isEmpty';
export * from './isFilePath';
export * from './isModuleName';
export * from './isObject';
export * from './isPlainObject';
export * from './toArray';
