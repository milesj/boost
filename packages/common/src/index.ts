/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import optimal, { Blueprint, Predicates } from 'optimal';
import Contract from './Contract';
import Path from './Path';
import PathResolver from './PathResolver';
import formatMs from './helpers/formatMs';
import instanceOf from './helpers/instanceOf';
import isEmpty from './helpers/isEmpty';
import isObject from './helpers/isObject';
import parseFile from './helpers/parseFile';
import requireModule from './helpers/requireModule';
import toArray from './helpers/toArray';

export * from './types';

export {
  Contract,
  Blueprint,
  Path,
  PathResolver,
  Predicates,
  formatMs,
  instanceOf,
  isEmpty,
  isObject,
  optimal,
  parseFile,
  requireModule,
  toArray,
};
