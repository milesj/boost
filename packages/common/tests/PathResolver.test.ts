import { getFixturePath, copyFixtureToNodeModule, getNodeModulePath } from '@boost/test-utils';
import { PathResolver, Path, LookupType } from '../src';

describe('PathResolver', () => {
  let resolver: PathResolver;

  beforeEach(() => {
    resolver = new PathResolver();
  });

  it('throws an error if no path is resolved', () => {
    resolver.lookupNodeModule('foo-bar');
    resolver.lookupFilePath('foo/bar');
    resolver.lookupFilePath('bar/baz');
    resolver.lookupNodeModule('bar-baz');

    expect(() => {
      resolver.resolvePath();
    }).toThrowErrorMatchingSnapshot();
  });

  it('returns first resolved node module or file path', () => {
    const cwd = getFixturePath('module-basic');

    resolver.lookupFilePath('qux.js', cwd); // Doesnt exist
    resolver.lookupNodeModule('@boost/unknown'); // Doesnt exist
    resolver.lookupFilePath('bar.js'); // Doesnt exist at this cwd
    resolver.lookupNodeModule('@boost/common'); // Exists

    expect(resolver.resolvePath()).toEqual(new Path(require.resolve('@boost/common')));
  });

  describe('file system', () => {
    it('returns first resolved path', () => {
      const cwd = getFixturePath('module-basic');

      resolver.lookupFilePath('qux.js', cwd); // Doesnt exist
      resolver.lookupFilePath('bar.js', cwd); // Exists
      resolver.lookupFilePath('foo.js', cwd); // Exists

      expect(resolver.resolve()).toEqual({
        lookupPath: new Path('bar.js'),
        resolvedPath: new Path(cwd, 'bar.js'),
        type: LookupType.FILE_SYSTEM,
      });
    });

    it('supports different cwds', () => {
      const cwd = getFixturePath('module-basic');

      resolver.lookupFilePath('qux.js', cwd); // Doesnt exist
      resolver.lookupFilePath('bar.js'); // Doesnt exist at this cwd
      resolver.lookupFilePath('foo.js', cwd); // Exists

      expect(resolver.resolvePath()).toEqual(new Path(cwd, 'foo.js'));
    });

    it('works with completely different parent folders and file extensions', () => {
      const src = new Path(__dirname, '../src');

      resolver.lookupFilePath('qux.js', getFixturePath('module-basic')); // Doesnt exist
      resolver.lookupFilePath('bar.js', src); // Doesnt exist
      resolver.lookupFilePath('toArray.ts', src.append('helpers')); // Exists

      expect(resolver.resolvePath()).toEqual(new Path(src, 'helpers/toArray.ts'));
    });
  });

  describe('node modules', () => {
    it('returns first resolved module', () => {
      resolver.lookupNodeModule('@boost/unknown'); // Doesnt exist
      resolver.lookupNodeModule('@boost/common'); // Exists
      resolver.lookupNodeModule('@boost/log'); // Exists

      expect(resolver.resolve()).toEqual({
        lookupPath: new Path('@boost/common'),
        resolvedPath: new Path(require.resolve('@boost/common')),
        type: LookupType.NODE_MODULE,
      });
    });

    it('works with sub-paths', () => {
      const unmock = copyFixtureToNodeModule('module-basic', 'test-module-path-resolver');

      resolver.lookupNodeModule('test-module-path-resolver/qux'); // Doesnt exist
      resolver.lookupNodeModule('test-module-path-resolver/bar'); // Exists (without extension)
      resolver.lookupNodeModule('test-module-path-resolver/foo.js'); // Exists

      expect(resolver.resolvePath()).toEqual(
        new Path(getNodeModulePath('test-module-path-resolver', 'bar.js')),
      );

      unmock();
    });
  });
});
