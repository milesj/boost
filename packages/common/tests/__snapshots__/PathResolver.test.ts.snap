// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathResolver throws an error if no path is resolved 1`] = `
"Failed to resolve a path using the following lookups (in order):
  - foo-bar (NODE_MODULE)
  - /Users/Miles/Projects/boost/foo/bar (FILE_SYSTEM)
  - /Users/Miles/Projects/boost/bar/baz (FILE_SYSTEM)
  - bar-baz (NODE_MODULE)"
`;
